### PR TITLE
[UI] Add a sign-in button in the app bar

### PR DESCRIPTION
**Pull request by iFlameing**
Related to issue #277 

@helfi92 we only have to remove the profile.displayName from the ListItem in usermenu.jsx and we achieve this feature. Currently ui looks when user is authenticated 
screenshot from 2019-02-21 21-54-44

and when the user is not authenticated looks like this

screenshot from 2019-02-21 21-54-55

How I can move this ListItem into seperate file. I have one thought i.e(we can make three separate component and call these component in usermenu.jsx.
1. If the user is not autenticated we can render this component
2. If the user is autenticated but for drawer
3. If the user is authnticated but for Appbar.
)

### CODE REVIEW

**Comment by iFlameing** [comment]
@helfi92  can you review this now I made the changes as you said :)

**Comment by iFlameing** [comment]
@helfi92 you can merge this now.

**Comment by iFlameing** [comment]
@helfi92 Do I need to do more modification or it is fine?

**Comment by helfi92** [comment]
> @helfi92 Do I need to do more modification or it is fine?

Sorry, I typically don't do work during the weekend. I haven't looked at it yet. I will look at it now :)

**Comment by iFlameing** [comment]
@helfi92 I fixed both of them you can merge this now.

**Comment by iFlameing** [comment]
@helfi92 I tried my best to refactor the code. can you review this or make some comment.

File: ui/src/components/UserMenu/index.jsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  handleClickSignOut = () => {
+    this.handleMenuClose();
+    this.props.onUnauthorize();
+    // Since Apollo caches query results, it’s important to get rid of them
+    // when the login state changes.
+    this.props.client.clearStore();
+  };
+
+  MenuClick = e => {
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Rename to `handleMenuClick`.

File: ui/src/components/UserMenu/index.jsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  },
+}))
+@withAuth
+@withApollo
+export default class UserMenu extends Component {
+  state = {
+    anchorEl: null,
+    signInDialogOpen: false,
+  };
+
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Define a `static propTypes` and `static defaultProps` methods. See the other component files for inspiration.

File: ui/src/components/UserMenu/index.jsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    return (
+      <Fragment>
+        {navOpen ? (
+          <UserMenuList
+            user={user}
+            signInDialogOpen={signInDialogOpen}
+            SignInDialogClosehandler={this.SignInDialogClose}
+            SignInDialogOpenhandler={this.SignInDialogOpen}
+            menuClickhandler={this.MenuClick}
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Rename to `onMenuClick={this.handleMenuClick}`.

File: ui/src/components/UserMenu/index.jsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+          />
+        ) : (
+          <UserMenuButton
+            user={user}
+            signInDialogOpen={signInDialogOpen}
+            SignInDialogClosehandler={this.SignInDialogClose}
+            SignInDialogOpenhandler={this.SignInDialogOpen}
+            menuClickhandler={this.MenuClick}
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: See comment above.

File: ui/src/components/UserMenu/index.jsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+          anchorEl={anchorEl}
+          open={Boolean(anchorEl)}
+          onClose={this.handleMenuClose}>
+          <MenuItem title="Your Profile" component={Link} to="/profile">
+            <AccountIcon className={classes.leftIcon} />
+            Account
+          </MenuItem>
+          <MenuItem
+            title={`Sign Out of ${process.env.APPLICATION_NAME}`}
+            onClick={this.handleClickSignOut}>
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Rename to `handleSignoutClick`.

File: ui/src/components/UserMenu/index.jsx (hunk 1 above)
Comment (iFlameing) [reply ui/src/components/UserMenu/index.jsx]: when I am renaming to handleMenuClick eslint give me an error 
` prop key to for handleMenuClick must begin with on`
what should I do?

Comment (iFlameing) [reply ui/src/components/UserMenu/index.jsx]: sorry I got it.

**Comment by iFlameing** [comment]
@helfi92 I made all the changes which are required.

File: ui/src/components/UserMenu/UserMenuButton.jsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    backgroundColor: theme.palette.secondary.main,
+  },
+  text: {
+    color: theme.palette.text.primary,
+    fontFamily: 'Roboto500',
+  },
+  icon: {
+    fill: theme.palette.common.white,
+    marginRight: theme.spacing.unit,
+    marginLeft: theme.spacing.unit,
```
Comment (helfi92) [inline ui/src/components/UserMenu/UserMenuButton.jsx]: Remove both `marginLeft` and `marginRight`. Add a marginLeft on the `UserMenuButton` instead. See other comment.

File: ui/src/components/UserMenu/index.jsx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    fill: theme.palette.text.primary,
+  },
+  leftIcon: {
+    marginRight: theme.spacing.unit,
+  },
+  username: {
+    overflow: 'hidden',
+    whiteSpace: 'nowrap',
+    textOverflow: 'ellipsis',
+  },
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Add a new class name:

```jsx
userMenuButton: {
  marginLeft: theme.spacing.unit,
},
```

File: ui/src/components/UserMenu/index.jsx [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+          <UserMenuList
+            user={user}
+            signInDialogOpen={signInDialogOpen}
+            onSignInDialogClose={this.handleSignInDialogClose}
+            onSignInDialogOpen={this.handleSignInDialogOpen}
+            onMenuClick={this.handleMenuClick}
+          />
+        ) : (
+          <UserMenuButton
+            user={user}
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Add `className={classes.userMenuButton}`. In order for the class to be applied, you will need to spread the props inside the `UserMenuButton` file. See other comment.

File: ui/src/components/UserMenu/UserMenuButton.jsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+@withApollo
+export default class UserMenuButton extends Component {
+  render() {
+    const {
+      classes,
+      user,
+      signInDialogOpen,
+      onSignInDialogOpen,
+      onSignInDialogClose,
+      onMenuClick,
```
Comment (helfi92) [inline ui/src/components/UserMenu/UserMenuButton.jsx]: Spread the rest of the props so we can give it to the button component:

```jsx
// ...
onMenuClick,
...props
```

File: ui/src/components/UserMenu/UserMenuButton.jsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    } = this.props;
+
+    if (!user) {
+      return (
+        <List component="nav">
+          <Button
+            variant="contained"
+            color="primary"
+            className={classes.icon}
+            onClick={onSignInDialogOpen}>
```
Comment (helfi92) [inline ui/src/components/UserMenu/UserMenuButton.jsx]: Add `{...props}` at the end so the class name can be applied.

File: ui/src/components/UserMenu/UserMenuButton.jsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+      onMenuClick,
+    } = this.props;
+
+    if (!user) {
+      return (
+        <List component="nav">
+          <Button
+            variant="contained"
+            color="primary"
+            className={classes.icon}
```
Comment (helfi92) [inline ui/src/components/UserMenu/UserMenuButton.jsx]: I don't think we need `classes.icon`. Remove this line.

File: ui/src/components/UserMenu/index.jsx [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+import Menu from '@material-ui/core/Menu';
+import MenuItem from '@material-ui/core/MenuItem';
+import AccountIcon from 'mdi-react/AccountIcon';
+import HandPeaceIcon from 'mdi-react/HandPeaceIcon';
+import { withAuth } from '../../utils/Auth';
+import UserMenuList from './UserMenuList';
+import UserMenuButton from './UserMenuButton';
+
+@withStyles(theme => ({
+  avatar: {
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: This is not being used. Remove unused classes from the rest.

**Comment by iFlameing** [comment]
@helfi92  is this possible to do because we want the same time to render userMenuList and userMenuButton from the userMenu component when docs is true.  can you tell me how I can proceed towards it?

**Comment by helfi92** [comment]
I'm not sure I understand the question but basically the docs site should just have the same behavior as the normal site, that is to use

screen shot 2019-02-27 at 11 00 13 am

for the left drawer. Let me know if that doesn't answer your question.

**Comment by iFlameing** [comment]
@helfi92 Mainly we have two components for showing signin button one UserMenuList(which is render when the left drawer is open) and other userMenuButton(which is rendered when a drawer is closed). 

The problem is in the documentation view where the drawer is permanent.

 In the permanent drawer, we want to render userMenuList component(in left side of taskcluster).

 And in appBar we want to render userMenuButton(on the left top) at the same time.

 But we can only render one component either userMenuButton or userMenuList at one time.

**Comment by iFlameing** [comment]
@helfi92 I resolve the problem I will make a commit tommorrow.

**Comment by helfi92** [comment]
> one UserMenuList(which is render when the left drawer is open)

Correct

> and other userMenuButton(which is rendered when a drawer is closed).

The userMenuButton has nothing to do with the left drawer. It should always be located top right in the app bar. If you open the the drawer, then we should be able to see both the UserMenuList in the drawer and UserMenuButton in the app bar.

> And in appBar we want to render userMenuButton(on the left top) at the same time

.Top right, not top left.

> But we can only render one component either userMenuButton or userMenuList at one time

Why is that?

**Comment by iFlameing** [comment]
@helfi92 I push a new commit with all the changes as you suggested above. And please make comment if something has to be modified :)

**Comment by iFlameing** [comment]
@helfi92 I fix the above issue.

File: ui/src/components/UserMenu/index.jsx [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  userMenuButton: {
+    marginLeft: theme.spacing.unit,
+  },
+}))
+@withAuth
+@withApollo
+export default class UserMenu extends Component {
+  static defaultProps = {
+    user: '',
+    navOpen: null,
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Remove `navOpen`.

File: ui/src/components/UserMenu/index.jsx [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    appBar: null,
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: `appBar: false,`

File: ui/src/components/UserMenu/index.jsx [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+  };
+
+  static propTypes = {
+    user: string,
+    navOpen: bool,
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Remove `navOpen`.

File: ui/src/components/UserMenu/index.jsx [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+  handleSignInDialogClose = () => {
+    this.setState({ signInDialogOpen: false });
+  };
+
+  handleSignInDialogOpen = () => {
+    this.setState({ signInDialogOpen: true });
+  };
+
+  render() {
+    const { classes, user, navOpen, appBar } = this.props;
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Remove `navOpen`.

File: ui/src/components/UserMenu/index.jsx [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    const { anchorEl, signInDialogOpen } = this.state;
+
+    return (
+      <Fragment>
+        {navOpen && appBar ? (
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Remove `navOpen`.

**Comment by iFlameing** [comment]
@helfi92 I modify the code as you suggested above. can you merge this now :).

thanks for all the help in creating pr.

**Comment by helfi92** [comment]
Looking at this now :)

**Comment by iFlameing** [comment]
@helfi92 I made all the above changes please take look!

**Comment by iFlameing** [comment]
@helfi92 I pushed the commit with the above modification :)

**Comment by iFlameing** [comment]
@helfi92 when I am rebasing it automatically resolve some code and change the code to previous commit . How I resolve this?

**Comment by helfi92** [comment]
When the rebase stop in the middle, you will need to resolve the change. Once that is resolved, you can continue with:

```bash
git add <file-changed>
git rebase --continue
```

**Comment by iFlameing** [comment]
@helfi92 I rebased all the pr :) sorry for package.lock.json I already updated my master earlier with serialize-error module which is missing when I am running web-server in master branch so I install it and for changing branch, I commited in the master. I don't why this happen. can you merge this now :)

File: services/web-server/package-lock.json [hunk 18]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (helfi92) [inline services/web-server/package-lock.json]: Remove this file.

**Comment by iFlameing** [comment]
@helfi92 How I can remove this. I don't know?

File: services/web-server/package-lock.json (hunk 18 above)
Comment (helfi92) [reply services/web-server/package-lock.json]:
```
rm <path/to/file>
git rm <path/to/file>
git commit -m 'Remove package-lock.json'
```

**Comment by helfi92** [comment]
https://github.com/taskcluster/taskcluster/pull/280#discussion_r261745582.

**Comment by iFlameing** [comment]
sorry for disturbing you so much. thanks for your help. I learned a lot from this conversation.

**Comment by iFlameing** [comment]
@helfi92 now you can merge this all test has been passed and all the conversation is resolved. thanks!

File: ui/src/components/Dashboard/index.jsx [hunk 19]
```diff
@@ -286,22 +286,18 @@ export default class Dashboard extends Component {
...
               <Tooltip placement="bottom" title="Page Information">
-                <IconButton
-                  onClick={this.handleHelpViewToggle}
-                  className={classes.appBarButton}>
+                <IconButton onClick={this.handleHelpViewToggle}>
                   <HelpIcon className={classes.appIcon} />
                 </IconButton>
               </Tooltip>
             )}
+            <UserMenu navOpen appBar />
```
Comment (helfi92) [inline ui/src/components/Dashboard/index.jsx]: Interesting. I thought we had removed `navOpen`?

File: ui/src/components/Dashboard/index.jsx [hunk 20]
```diff
@@ -243,10 +243,10 @@ export default class Dashboard extends Component {
           </Typography>
         </div>
         <Divider />
-        <UserMenu />
+        {docs ? <UserMenu /> : <UserMenu navOpen />}
```
Comment (helfi92) [inline ui/src/components/Dashboard/index.jsx]: Is `navOpen` being used?

File: ui/src/components/Dashboard/index.jsx (hunk 19 above)
Comment (iFlameing) [reply ui/src/components/Dashboard/index.jsx]: This problem occur when I have done rebase. It automatically got resolve adding this.

File: ui/src/components/Dashboard/index.jsx (hunk 20 above)
Comment (iFlameing) [reply ui/src/components/Dashboard/index.jsx]: yes it is being used here you can refer previous commit  https://github.com/taskcluster/taskcluster/pull/280/commits/2f8194b1e4eb994e9bd613096cb77d997b682414

File: ui/src/components/Dashboard/index.jsx (hunk 19 above)
Comment (iFlameing) [reply ui/src/components/Dashboard/index.jsx]: No, navOpen is being used in the code. see the below comment I put the reference of the code.

**Comment by iFlameing** [comment]
@helfi92 I think it is still not solved because after adding your changes the right avatar not position correctly in the center. please check it out.
screenshot from 2019-03-02 04-49-03

**Comment by helfi92** [comment]
I have to sign off now but I'll be back in a week to hopefully merge this :) I just don't want to merge something right now and not be available if anything goes wrong while I'm away. Sorry for that. Awesome work!

**Comment by iFlameing** [comment]
@helfi92 I think rebase done some terrible work it resolve some file automatically which lead to this problem. I really tried to make this merge today but I think I can close this pr and make a new pr with all the change you suggested me in this pr. so that it is easy to review and then we will merge it.

**Comment by helfi92** [comment]
Works with me :)

**Comment by iFlameing** [comment]
@helfi92 It works in firefox but not in chrome.

**Comment by iFlameing** [comment]
@helfi92 It is the problem related to padding in chrome.

**Comment by djmitche** [comment]
@iFlameing @helfi92 is away this week.  I'm less familiar with this code, but can do a review when you believe it is ready to merge.  As for differences between browsers -- I bet you know more than i do about that, so I will leave you to puzzle it out.  Let me know when you'd like a final review.

**Comment by iFlameing** [comment]
@djmitche I think it is ready for final review. I will open a new pr for the browser issue when this is merge.
thanks!

File: ui/src/components/Dashboard/index.jsx (hunk 20 above)
Comment (helfi92) [reply ui/src/components/Dashboard/index.jsx]: I still don't see how it's being used. What is it doing? Didn't we remove it in https://github.com/taskcluster/taskcluster/pull/280#discussion_r261261456?

File: ui/src/components/Dashboard/index.jsx (hunk 19 above)
Comment (iFlameing) [reply ui/src/components/Dashboard/index.jsx]: Going to investigate and if it is not needed I will remove it.

File: ui/src/components/Dashboard/index.jsx (hunk 20 above)
Comment (helfi92) [reply ui/src/components/Dashboard/index.jsx]: Replace this line with `<UserMenu />`.

File: ui/src/components/Dashboard/index.jsx (hunk 19 above)
Comment (helfi92) [reply ui/src/components/Dashboard/index.jsx]: Remove `navOpen`.

File: ui/src/components/UserMenu/UserMenuButton.jsx [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+import SignInDialog from '../SignInDialog';
+import { withAuth } from '../../utils/Auth';
+
+@withStyles(theme => ({
+  icon: {
+    fill: theme.palette.common.white,
+  },
+  avatarButton: {
+    height: 6 * theme.spacing.unit,
+    width: 6 * theme.spacing.unit,
```
Comment (helfi92) [inline ui/src/components/UserMenu/UserMenuButton.jsx]: Add `padding: 0` to center the button.

File: ui/src/components/UserMenu/UserMenuButton.jsx [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+import IconButton from '@material-ui/core/IconButton';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import AccountCircleIcon from 'mdi-react/AccountCircleIcon';
+import Button from '../Button';
+import SignInDialog from '../SignInDialog';
+import { withAuth } from '../../utils/Auth';
+
+@withStyles(theme => ({
+  icon: {
+    fill: theme.palette.common.white,
```
Comment (helfi92) [inline ui/src/components/UserMenu/UserMenuButton.jsx]: Is this having any effect? I think we can remove it.

Comment (iFlameing) [reply ui/src/components/UserMenu/UserMenuButton.jsx]: yes, we can remove this.

**Comment by iFlameing** [comment]
@helfi92 I made the changes as you said and also the padding issue is solved in chrome browser :)

File: ui/src/components/UserMenu/UserMenuButton.jsx [hunk 23]
```diff
@@ -12,12 +12,10 @@ import SignInDialog from '../SignInDialog';
 import { withAuth } from '../../utils/Auth';
 
 @withStyles(theme => ({
-  icon: {
-    fill: theme.palette.common.white,
-  },
   avatarButton: {
     height: 6 * theme.spacing.unit,
     width: 6 * theme.spacing.unit,
+    padding: '0',
```
Comment (helfi92) [inline ui/src/components/UserMenu/UserMenuButton.jsx]: Remove quotations. `padding: 0`.

Comment (iFlameing) [reply ui/src/components/UserMenu/UserMenuButton.jsx]: I removed the quotations.

File: ui/src/components/UserMenu/index.jsx [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+    '& svg': {
+      fill: THEME.PRIMARY_TEXT_DARK,
+    },
+    '&:hover': {
+      backgroundColor: darken(THEME.PRIMARY_DARK, 0.5),
+    },
+  },
+}))
+@withAuth
+@withApollo
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]: Let's make `@withAuth` and `@withApollo` come before `@withStyles`.

File: ui/src/components/UserMenu/index.jsx [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+      <Fragment>
+        {appBar ? (
+          <UserMenuButton
+            avatarProps={{
+              className: classes.buttonAvatar,
+            }}
+            buttonProps={{
+              classes: { containedPrimary: classes.buttonContainedPrimary },
+            }}
+            className={classes.userMenuButton}
```
Comment (helfi92) [inline ui/src/components/UserMenu/index.jsx]:
```jsx
import classNames from 'classnames';

// ...

className={classNames({ [classes.userMenuButton]: !user })}
```

**Comment by iFlameing** [comment]
@helfi92 can you merge this. I pushed the modify changes :)

**Comment by iFlameing** [comment]
@helfi92 Adding these changes remove this problem. please! take a look.

--- a/ui/src/components/UserMenu/UserMenuButton.jsx
+++ b/ui/src/components/UserMenu/UserMenuButton.jsx
@@ -19,6 +19,9 @@ import { withAuth } from '../../utils/Auth';
     width: 6 * theme.spacing.unit,
     padding: 0,
   },
+  Button: {
+    whiteSpace: 'nowrap',
+  },
 }))
 export default class UserMenuButton extends Component {
   static propTypes = {
@@ -49,7 +52,7 @@ export default class UserMenuButton extends Component {
       return (
         <Fragment>
           <Button
-            className={className}
+            className={classNames({ [classes.Button]: !user })}
             variant="contained"
             color="primary"
             onClick={onSignInDialogOpen}

**Comment by iFlameing** [comment]
@helfi92 I like the changes and pushed the modify code :)

**Comment by helfi92** [comment]
Awesome, I will take a look soon!

**Pull request merged**